src/mocks.ts

```diff
@@ -133,16 +76,168 @@ export const reduceAllMocksForScenario = (
+
+    return result;
+  }, {});
+  const httpMocks = Object.values(httpMocksByUrlAndMethod);
+
+  const graphQlMocksByUrlAndOperations = initialGraphQlMocks.reduce<
+    Record<string, Record<string, Operation>>
+  >((result, mock) => {
+    const { url, operations } = mock;
+
+    const operationsByName: Record<string, Operation> = result[url.toString()]
+      ? result[url.toString()]
+      : {};
+
+    operations.forEach(operation => {
+      // Always take the latest operation
+      operationsByName[operation.operationName] = operation;
+    });
+
+    result[url.toString()] = operationsByName;
+    return result;
+  }, {});
+  const graphQlMocks = Object.entries(graphQlMocksByUrlAndOperations).map(
+    ([url, operationsByName]) => {
+      return {
+        method: 'GRAPHQL',
+        url: RegExp(url.replace(/^\/(.*)\/$/, '$1')),
+        operations: Object.values(operationsByName)
+      };
+    }
+  ) as GraphQLMock[];
+
+  return (httpMocks as any).concat(graphQlMocks);
 };
 
+/**
+ * Mocks the right HTTP method for a REST mock.
+ */
+function handleRestMock({
+  method,
+  url,
+  response,
+  responseCode = 200,
+  responseHeaders,
+  delay = 0
+}: HttpMock) {
+  const finalResponse = {
+    body: response,
+    status: responseCode,
+    headers: responseHeaders
+  };
+
+  switch (method) {
+    case 'GET':
+      FetchMock.get(url, () => addDelay(delay).then(() => finalResponse), {
+        overwriteRoutes: false
+      });
+      XHRMock.get(url, xhrMockDelay(finalResponse, delay));
+      break;
+    case 'POST':
+      FetchMock.post(url, () => addDelay(delay).then(() => finalResponse), {
+        overwriteRoutes: false
+      });
+      XHRMock.post(url, xhrMockDelay(finalResponse, delay));
+      break;
+    case 'PUT':
+      FetchMock.put(url, () => addDelay(delay).then(() => finalResponse), {
+        overwriteRoutes: false
+      });
+      XHRMock.put(url, xhrMockDelay(finalResponse, delay));
+      break;
+    case 'DELETE':
+      FetchMock.delete(url, () => addDelay(delay).then(() => finalResponse), {
+        overwriteRoutes: false
+      });
+      XHRMock.delete(url, xhrMockDelay(finalResponse, delay));
+      break;
+    default:
+      throw new Error(
+        `Unrecognised HTTP method ${method} - please check your mock configuration`
+      );
+  }
+}
+
+/**
+ * Mocks the right HTTP method for a GraphQL mock.
+ */
+function handleGraphQLMock({ url, operations }: GraphQLMock) {
+  const graphQLErrorResponse = { errors: [] };
+  const findMockGet = (u: string) => {
+    const parsedUrl = new URL(u);
+    const operationName = parsedUrl.searchParams.get('operationName');
+    return operations.find(o => o.operationName === operationName);
+  };
+  const findMockPost = postBody => {
+    const jsonBody =
+      typeof postBody === 'string' ? JSON.parse(postBody) : postBody;
+    return operations.find(
+      ({ operationName }) => operationName === jsonBody.operationName
```

Comment:
Specifying an `operationName` is optional in GraphQL and could bundle a few queries into one operation - you should probably add to the docs that operation names are required within the app codebase as well as the mocks

Reply:
Added this into the gotchas section

Have you got any ideas about how we could get around this?

Reply:
There's no real way round this. The HTTP requests were able to be targeted, because they were unique by their url and method. With GraphQL, given it always goes to the same url, the only way of targeting it specifically is by looking at the operation name.

This can be done in GraphQL:

```
query A { ... }
query B { ... }
```
In this example the GraphQL client has to send up the `operationName` in the body, because it has no way of knowing which query to execute. However in the following:

```
query A { ... }
```

...the GraphQL client doesn't need to send up the operation name, because it's in the query and can be parsed.

https://github.com/kenneth-gray/data-mocks-server/blob/v5.6.0/src/index.ts#L548-L576 is an example of looking out for the different ways of getting `query`, `variables` and `operationName` from a request. I know this is working for `urql`, but it's probably worth testing with `apollo` (GraphQL) just in case.

I think it's fine to ask users of this library to add and operationName to everything. We probably want to encourage users to think about the long term and having operation names means they can think about adding monitoring at some point.

The only other way to solve this problem would be to go to a lower level: the resolvers. I wonder what kind of maintenance overhead this would be in comparison to targeting entire queries :thinking: :thinking:.

Reply:
> The only other way to solve this problem would be to go to a lower level: the resolvers. I wonder what kind of maintenance overhead this would be in comparison to targeting entire queries :thinking: :thinking:.

This would probably get messy quickly, because you could end up trying to implement a real GraphQL server at this point!

Reply:
Yeah, I think @kevrowe made some comment alluding to this in a Slack conversation I had with him.

I think, for now, let's go with requiring operation name to be included and see if we get any feedback on that.

I'm a bit of a GraphQL noob, so I'm not sure what the common convention is regarding adding operation names to queries, but to me, it seems quite logical and you have a point re. being useful for monitoring @kenneth-gray